### PR TITLE
[rllib] PyTorch Models for A3C

### PR DESCRIPTION
**Pull request by richardliaw**
- [x] May need to refactor into modelcatalog

Currently only supports discrete control.

- [x] testing comparison
- [x] Currently, policy is strict in limiting thread access - check if not needed

Note: Without locking, pytorch multithreaded loading/setting/accessing state will can cause segfaults or corrupt data structures. 

Performance notes: For PongDeterministic-v4 on VisionNet, to reach iteration score > 15:
  - pytorch with lock - 23839 sec (397 min)
  - tensorflow - 15346 sec (256 min)

cc @ericl

### CODE REVIEW

File: python/ray/rllib/a3c/a3c.py [hunk 1]
```diff
@@ -20,10 +21,11 @@
     "num_batches_per_iteration": 100,
     "batch_size": 10,
     "use_lstm": True,
+    "use_pytorch": False,
     "model": {"grayscale": True,
               "zero_mean": False,
               "dim": 42,
-              "channel_major": True}
+              "channel_major": False}
```
Comment (richardliaw) [inline python/ray/rllib/a3c/a3c.py]: somehow this was defaulted incorrectly

File: python/ray/rllib/models/preprocessors.py [hunk 2]
```diff
@@ -30,14 +30,14 @@ def _init(self):
         self._grayscale = self._options.get("grayscale", False)
         self._zero_mean = self._options.get("zero_mean", True)
         self._dim = self._options.get("dim", 80)
-        self._pytorch = self._options.get("pytorch", False)
+        self._channel_major = self._options.get("channel_major", False)
```
Comment (richardliaw) [inline python/ray/rllib/models/preprocessors.py]: this was also an incorrect residual

File: python/ray/rllib/models/pytorch/visionnet.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    def hidden_layers(self, inputs):
+        """ Internal method - pass in Variables, not numpy arrays
+
+        args:
+            inputs: observations and features"""
+        res = self._convs(inputs)
+        res = res.view(-1, 32*8*8)
+        return self.fc1(res)
+
+    def forward(self, inputs):
```
Comment (ericl) [inline python/ray/rllib/models/pytorch/visionnet.py]: Should these be documented in the base model class?

**Review by ericl — COMMENTED** [review]
Cool, so what's next? have you run any comparisons with the tensorflow model?

File: python/ray/rllib/models/pytorch/model.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import torch.nn as nn
+
+
+class Model(nn.Module):
```
[the remark this answers is not in the document]
Comment (richardliaw) [reply python/ray/rllib/models/pytorch/model.py]: Done.

**Comment by richardliaw** [comment]
Currently running a benchmark task on EC2 for both Linear and Conv models.

Next immediate steps are to (localized to the A3C/ModelCatalog):

1. Introduce the LSTM for PyTorch
2. Investigate and provide GPU support for PyTorch

**Review by ericl — CHANGES_REQUESTED** [review]
I think we should go ahead and merge this. Couple thing first though:

- Could we add a jenkins test that runs the pytorch a3c for a couple iterations?
- It would be nice if RLlib didn't acquire a hard dependency on pytorch. Can you make sure that things still run without torch installed (perhaps we should leave torch uninstalled on travis for this reason, and do a simple import test on rllib and check it doesn't crash?)

**Review by ericl — APPROVED** [review]

**Pull request merged**